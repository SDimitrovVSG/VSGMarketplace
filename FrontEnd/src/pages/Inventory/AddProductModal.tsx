import {
    TextField,
    FormControl,
    InputLabel,
    Select,
    MenuItem,
    CircularProgress,
    FormHelperText,
} from '@mui/material';
import { ICategory, IFormInputs, ILocation, IProduct } from '../../types/types.ts';
import { useState, Dispatch, SetStateAction } from 'react';
import { useCreateProductMutation } from '../../services/productsService.ts';
import { useCreateImageMutation } from '../../services/imageService.ts';
import { useGetCategoriesQuery } from '../../services/categoriesService.ts';
import { useGetLocationQuery } from '../../services/locationsService.ts';
import { imagePlaceholder } from '../../utils/imagePlaceholder.ts';
import { uploadImage } from '../../utils/uploadImage.ts';
import { useForm } from 'react-hook-form';
import { toast } from 'react-toastify';
import Modal from '../../components/Modal.tsx';

type AddModalProps = {
    products: IProduct[] | undefined;
    setProducts: Dispatch<SetStateAction<IProduct[]>>;
    showAddModal: boolean;
    setShowAddModal: Dispatch<SetStateAction<boolean>>;
};

const AddProductModal = ({
    //products,
    setProducts,
    showAddModal,
    setShowAddModal,
}: AddModalProps) => {
    const [createProduct] = useCreateProductMutation();
    const [createImage] = useCreateImageMutation();
    const { data: categories } = useGetCategoriesQuery();
    const { data: locations } = useGetLocationQuery();
    const [imageUrl, setImageUrl] = useState(imagePlaceholder);
    const [selectOption, setSelectOption] = useState('');
    const [locationOption, setLocationOption] = useState('');
    const {
        register,
        handleSubmit,
        formState: { errors, isSubmitting },
        getValues,
        reset,
        watch
    } = useForm<IFormInputs>({
        defaultValues: {
            code: '',
            fullName: '',
            description: '',
            categoryId: null,
            locationId: null,
            quantityForSale: null,
            price: null,
            quantity: null,
            image: null,
        },
        mode: 'onChange',
    });

    const onSubmit = async (data: IFormInputs): Promise<void> => {
        const image: { name: string } | File = data.image ? data.image[0] : { name: '' };
        data.quantityForSale = data.quantityForSale || null;
        data.price = data.price || null;

        const response = await createProduct(data) as { data: IProduct };
        if ('error' in response) {
            setShowAddModal(false);
            return;
        }

        const responseData = response.data;
        const selectedCategory = categories?.filter(c => data.categoryId === c.id)[0] as ICategory;
        const selectedCity = locations?.filter(l => data.locationId === l.id)[0] as ILocation;
        const id = responseData.id;

        if (image.name) {
            const imageForm = new FormData();
            imageForm.set('picture', image as File);
            const imgRes = await createImage({ id, imageForm }) as { data: string };
            const newImgUrl = imgRes.data;
            const newProduct = { ...data, id: id, type: selectedCategory.type, city: selectedCity.city, imageUrl: newImgUrl } as IProduct;
            setProducts((oldProducts) => [...oldProducts, newProduct]);
        } else {
            const newProduct = { ...data, id: id, type: selectedCategory.type, city: selectedCity.city } as IProduct;
            setProducts((oldProducts) => [...oldProducts, newProduct]);
        }

        setSelectOption('');
        setLocationOption('');
        toast.success('Created successfully!');
        setShowAddModal(false);
        setImageUrl(imagePlaceholder);
        reset();
    };

    return (
        <Modal showModal={showAddModal} setShowModal={setShowAddModal}>
            <form className='modalContent' onSubmit={handleSubmit(onSubmit)}>
                <div className='row'>
                    <div className='leftModal'>
                        <h2>Add New Item</h2>
                        <TextField
                            className='formInput'
                            type='text'
                            label='Code *'
                            variant='standard'
                            error={Boolean(errors.code)}
                            helperText={errors.code?.message}
                            {...register('code', {
                                required: 'Code field is required',
                                // validate: value => products ? products.some(p => p.code !== value) : false || 'Code already exists'
                            })}
                        />
                        <TextField
                            className='formInput'
                            type='text'
                            label='Name *'
                            variant='standard'
                            error={Boolean(errors.fullName)}
                            helperText={errors.fullName?.message}
                            {...register('fullName', { required: 'Name field is required' })}
                        />
                        <TextField
                            className='description'
                            label='Description'
                            multiline
                            rows={4}
                            variant='standard'
                            {...register('description')}
                        />
                        <FormControl variant='standard' className='formInput'>
                            <InputLabel focused={false}>Category *</InputLabel>
                            <Select
                                value={selectOption}
                                error={(Boolean(errors.categoryId) && Boolean(errors.categoryId?.message))}
                                {...register('categoryId', {
                                    onChange: (e) => setSelectOption(e.target.value),
                                    required: 'Category field is required',
                                })}
                            >
                                {categories?.map((c: ICategory) => (
                                    <MenuItem value={c.id} key={c.id}>
                                        {c.type}
                                    </MenuItem>
                                ))}
                            </Select>
                            <FormHelperText error>
                                {watch('categoryId') === null && errors.categoryId?.message ? errors.categoryId?.message : ''}
                            </FormHelperText>
                        </FormControl>
                        <FormControl variant='standard' className='formInput'>
                            <InputLabel focused={false}>Location *</InputLabel>
                            <Select
                                value={locationOption}
                                error={Boolean(errors.locationId)}
                                {...register('locationId', {
                                    onChange: (e) => setLocationOption(e.target.value),
                                    required: 'Location field is required',
                                })}
                            >
                                {locations?.map((l: ILocation) => (
                                    <MenuItem value={l.id} key={l.id}>
                                        {l.city}
                                    </MenuItem>
                                ))}
                            </Select>
                            <FormHelperText error>
                                {watch('locationId') === null && errors.locationId?.message ? errors.locationId?.message : ''}
                            </FormHelperText>
                        </FormControl>
                        <TextField
                            className='formInput'
                            type='number'
                            label='Qty For Sale'
                            variant='standard'
                            error={Boolean(errors.quantityForSale)}
                            helperText={errors.quantityForSale?.message}
                            {...register('quantityForSale', {
                                min: {
                                    value: 0,
                                    message: 'You cannot input less than 0'
                                },
                            })}
                        />
                        <TextField
                            className='formInput'
                            type='number'
                            label='Sale Price'
                            variant='standard'
                            error={Boolean(errors.price)}
                            helperText={errors.price?.message}
                            {...register('price', {
                                min: {
                                    value: 0,
                                    message: 'You cannot input less than 0'
                                }
                            })}
                        />
                        <TextField
                            className='formInput'
                            type='number'
                            label='Qty *'
                            variant='standard'
                            error={Boolean(errors.quantity)}
                            helperText={errors.quantity?.message}
                            {...register('quantity', {
                                required: 'Quantity field is required',
                                min: {
                                    value: 0,
                                    message: 'Quantity must be 0 or higher'
                                },
                                validate: value => value as number >= Number(getValues('quantityForSale')) || 'Qty For Sale cannot be higher than Qty'
                            })}
                        />
                    </div>
                    <div className='rightModal'>
                        <img className='currentImg' src={imageUrl} />
                        <input
                            id='uploadInput'
                            className='inputImage'
                            accept='image/*'
                            type='file'
                            {...register('image', {
                                onChange: (e) => setImageUrl(uploadImage(e))
                            })}
                        />
                        <div className='uploadDelete'>
                            <label htmlFor='uploadInput' className='uploadImg'>
                                Upload
                            </label>
                            <button type='button' className='deleteImg' onClick={() => setImageUrl(imagePlaceholder)}>Remove</button>
                        </div>
                    </div>
                </div>
                {isSubmitting
                    ? <CircularProgress className='circular' />
                    : <button type='submit'>Add</button>
                }
            </form>
        </Modal>
    );
};

export default AddProductModal;
