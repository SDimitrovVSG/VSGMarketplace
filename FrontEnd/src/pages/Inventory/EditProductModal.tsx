import { useDeleteImageMutation, useEditImageMutation } from '../../services/imageService.ts';
import { useState, Dispatch, SetStateAction, } from 'react';
import { IFormInputs, ICategory, IProduct, ILocation } from '../../types/types.ts';
import { useEditProductMutation } from '../../services/productsService.ts';
import { useGetCategoriesQuery } from '../../services/categoriesService.ts';
import { useGetLocationQuery } from '../../services/locationsService.ts';
import { imagePlaceholder } from '../../utils/imagePlaceholder.ts';
import { uploadImage } from '../../utils/uploadImage.ts';
import { useForm } from 'react-hook-form';
import { toast } from 'react-toastify';
import Modal from '../../components/Modal.tsx';
import {
    TextField,
    FormControl,
    InputLabel,
    Select,
    MenuItem,
    FormHelperText,
} from '@mui/material';

type EditModalProps = {
    setProducts: Dispatch<SetStateAction<IProduct[]>>;
    product: IProduct;
    showEditModal: boolean;
    setShowEditModal: Dispatch<SetStateAction<boolean>>;
};

const EditProductModal = ({ setProducts, product, showEditModal, setShowEditModal }: EditModalProps) => {
    const [locationOption, setLocationOption] = useState(product.locationId.toString());
    const [selectOption, setSelectOption] = useState(product.categoryId.toString());
    const [imageUrl, setImageUrl] = useState(product.imageUrl || imagePlaceholder);
    const [editProduct, { isLoading: fetchingProduct }] = useEditProductMutation();
    const [editImage, { isLoading: fetchingImage }] = useEditImageMutation();
    const [deleteImage] = useDeleteImageMutation();
    const { data: categories } = useGetCategoriesQuery();
    const { data: locations } = useGetLocationQuery();
    const {
        register,
        handleSubmit,
        formState: { errors },
        getValues,
    } = useForm<IFormInputs>({
        defaultValues: {
            code: product.code,
            fullName: product.fullName,
            description: product.description,
            categoryId: product.categoryId,
            locationId: product.locationId,
            quantityForSale: product.quantityForSale || null,
            price: product.price || null,
            quantity: product.quantity,
            image: null
        },
        mode: 'all',
    });

    const onSubmit = async (data: IFormInputs) => {
        const image: { name: string } | File = data.image ? data.image[0] : { name: '' };
        data.quantityForSale = data.quantityForSale || null;
        data.price = data.price || null;

        const currentImg = document.querySelector('.currentImg') as HTMLImageElement;
        const selectedCategory = categories?.filter(c => data.categoryId === c.id)[0] as ICategory;
        const selectedCity = locations?.filter(l => data.locationId === l.id)[0] as ILocation;
        const id = product.id;

        if (image.name) {
            const imageForm = new FormData();
            imageForm.append('newPicture', image as File);
            const imgRes = await editImage({ id, imageForm }) as { data: string };
            const newImgUrl = imgRes.data;
            setImageUrl(newImgUrl);
            const editedProduct = { ...data, id: id, type: selectedCategory.type, city: selectedCity.city, imageUrl: newImgUrl } as IProduct;
            setProducts(oldProducts => oldProducts.map(p => p.id === editedProduct.id ? editedProduct : p));
        } else if (currentImg.src !== product.imageUrl && product.imageUrl !== null) {
            await deleteImage(id);
        }

        const response = await editProduct({ id, data }) as { data: IProduct };

        if ('error' in response) {
            setShowEditModal(false);
            return
        } else {
            toast.success('Modified successfully!');
        }

        const editedProduct = { ...data, id: id, type: selectedCategory.type, city: selectedCity.city } as IProduct;
        setProducts(oldProducts => oldProducts.map(p => p.id === editedProduct.id ? editedProduct : p));
        setShowEditModal(false);
    };

    return (
        <Modal showModal={showEditModal} setShowModal={setShowEditModal} >
            <form className='editForm' onSubmit={handleSubmit(onSubmit)}>
                <div className='row'>
                    <div className='leftModal'>
                        <h2>Modify Item</h2>
                        <TextField
                            className='formInput'
                            type='text'
                            label='Code *'
                            variant='standard'
                            error={Boolean(errors.code)}
                            helperText={errors.code?.message}
                            {...register('code', { required: 'Code field is required' })}
                        />
                        <TextField
                            className='formInput'
                            type='text'
                            label='Name *'
                            variant='standard'
                            error={Boolean(errors.fullName)}
                            helperText={errors.fullName?.message}
                            {...register('fullName', { required: 'Name field is required' })}
                        />
                        <TextField
                            className='description'
                            label='Description'
                            multiline
                            rows={4}
                            variant='standard'
                            {...register('description')}
                        />
                        <FormControl variant='standard' className='formInput'>
                            <InputLabel focused={false}>Category *</InputLabel>
                            <Select
                                value={selectOption}
                                error={Boolean(errors.categoryId)}
                                {...register('categoryId', {
                                    onChange: (e) => setSelectOption(e.target.value),
                                    required: 'Category field is required',
                                })}
                            >
                                <MenuItem value={selectOption} key={selectOption}>
                                    {product.type}
                                </MenuItem>
                                {categories?.map((c: ICategory) => (
                                    c.type !== product.type &&
                                    <MenuItem value={c.id} key={c.id}>
                                        {c.type}
                                    </MenuItem>
                                ))}
                            </Select>
                        </FormControl>
                        <FormControl variant='standard' className='formInput'>
                            <InputLabel focused={false}>Location *</InputLabel>
                            <Select
                                value={locationOption}
                                error={Boolean(errors.locationId)}
                                {...register('locationId', {
                                    onChange: (e) => setLocationOption(e.target.value),
                                    required: 'Location field is required',
                                })}
                            >
                                <MenuItem value={locationOption} key={locationOption}>
                                    {product.city}
                                </MenuItem>
                                {locations?.map((l: ILocation) => (
                                    l.city !== product.city &&
                                    <MenuItem value={l.id} key={l.id}>
                                        {l.city}
                                    </MenuItem>
                                ))}
                            </Select>
                            <FormHelperText error>
                                {errors.locationId?.message}
                            </FormHelperText>
                        </FormControl>
                        <TextField
                            className='formInput'
                            type='number'
                            label='Qty For Sale'
                            variant='standard'
                            error={Boolean(errors.quantityForSale)}
                            helperText={errors.quantityForSale?.message}
                            {...register('quantityForSale', {
                                min: {
                                    value: 0,
                                    message: 'You cannot input less than 0'
                                },
                            })}
                        />
                        <TextField
                            className='formInput'
                            type='number'
                            label='Sale Price'
                            variant='standard'
                            {...register('price', {
                                min: {
                                    value: 0,
                                    message: 'You cannot input less than 0'
                                }
                            })}
                        />
                        <TextField
                            className='formInput'
                            type='number'
                            label='Qty *'
                            variant='standard'
                            error={Boolean(errors.quantity)}
                            helperText={errors.quantity?.message}
                            {...register('quantity', {
                                required: 'Quantity field is required',
                                min: {
                                    value: 0,
                                    message: 'Quantity must be 0 or higher'
                                },
                                validate: value => value as number >= Number(getValues('quantityForSale')) || 'Qty cannot be less than Qty For Sale'
                            })}
                        />
                    </div>
                    <div className='rightModal'>
                        <img className='currentImg' src={imageUrl} />
                        <input
                            id='editUploadInput'
                            className='inputImage'
                            accept='image/*'
                            type='file'
                            {...register('image', {
                                onChange: (e) => setImageUrl(uploadImage(e))
                            })}
                        />
                        <div className='uploadDelete'>
                            <label htmlFor='editUploadInput' className='uploadImg'>
                                Upload
                            </label>
                            <button type='button' className='deleteImg' onClick={() => setImageUrl(imagePlaceholder)}>Remove</button>
                        </div>
                    </div>
                </div>
                <button type='submit' disabled={fetchingProduct || fetchingImage}>{(fetchingProduct || fetchingImage) ? 'Submitting...' : 'Modify'}</button>
            </form>
        </Modal>
    );
};


export default EditProductModal;
