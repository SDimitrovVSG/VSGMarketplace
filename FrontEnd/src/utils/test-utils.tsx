import { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import store from '../redux/store';
import { BrowserRouter } from 'react-router-dom';

type ProvidersProps = {
  children: ReactElement;
};

const AllTheProviders = ({ children }: ProvidersProps) => {
  return (
    <Provider store={store}>
      <BrowserRouter>
        {children}
      </BrowserRouter>
    </Provider>
  )
}

const customRender = (ui: ReactElement, options?: object) => (
  render(ui, { wrapper: AllTheProviders, ...options })
);

export * from '@testing-library/react';

// override render method
export { customRender as render }