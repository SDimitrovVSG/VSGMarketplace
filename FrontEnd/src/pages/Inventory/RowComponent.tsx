import { TableCell, TableRow, tableCellClasses, styled, Fade } from '@mui/material';
import { useState, useRef, Dispatch, SetStateAction } from 'react';
import { useDeleteProductMutation } from '../../services/productsService.ts';
import { IProduct } from '../../types/types.ts';
import { toast } from 'react-toastify';
import DeleteOutlineIcon from '@mui/icons-material/DeleteOutline';
import EditProductModal from './EditProductModal.tsx';
import EditIcon from '@mui/icons-material/Edit';
import Popup from '../../components/Popup.tsx';

const StyledTableCell = styled(TableCell)(({ theme }) => ({
    [`&.${tableCellClasses.head}`]: {
        backgroundColor: theme.palette.common.black,
        color: theme.palette.common.white,
    },
    [`&.${tableCellClasses.body}`]: {
        fontSize: 14,
    },
}));

const StyledTableRow = styled(TableRow)(({ theme }) => ({
    '&:nth-of-type(odd)': {
        backgroundColor: theme.palette.action.hover,
    },
    '&:last-child StyledTableCell, &:last-child th': {
        border: 0,
    },
}));

type RowComponentProps = {
    product: IProduct;
    setProducts: Dispatch<SetStateAction<IProduct[]>>
};

const RowComponent = ({ product, setProducts }: RowComponentProps) => {
    const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
    const [showEditModal, setShowEditModal] = useState(false);
    const [deleteProduct] = useDeleteProductMutation();
    const rowRef = useRef<HTMLTableRowElement>(null);

    const handleDelete = async () => {
        setAnchorEl(null);
        const response = await deleteProduct(product.id);
        if ('error' in response) {
            return;
        }

        toast.success('Deleted successfully!');
        setProducts(oldProducts => oldProducts.filter(p => p !== product));
    }

    const popupMessage = `Are you sure you want to remove this item ?`;

    return (
        <>
            <EditProductModal setProducts={setProducts} product={product} showEditModal={showEditModal} setShowEditModal={setShowEditModal} />
            <Popup popupMessage={popupMessage} onYes={handleDelete} anchorEl={anchorEl} setAnchorEl={setAnchorEl} />
            <Fade in={true} timeout={500}>
                <StyledTableRow ref={rowRef} id={product.id.toString()}>
                    <StyledTableCell>{product.code}</StyledTableCell>
                    <StyledTableCell>{product.fullName}</StyledTableCell>
                    <StyledTableCell>{product.type}</StyledTableCell>
                    <StyledTableCell>{product.quantityForSale || 0}</StyledTableCell>
                    <StyledTableCell>{product.quantity}</StyledTableCell>
                    <StyledTableCell>
                        <a className='edit' onClick={() => setShowEditModal(true)}>
                            <EditIcon />
                        </a>
                        <a className='delete' onClick={e => setAnchorEl(e.currentTarget)}>
                            <DeleteOutlineIcon />
                        </a>
                    </StyledTableCell>
                </StyledTableRow>
            </Fade>
        </>
    );
};

export default RowComponent;
