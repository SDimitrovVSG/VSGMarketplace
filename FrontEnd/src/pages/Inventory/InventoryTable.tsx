import { Table, TableBody, TableCell, TableContainer, TableHead, TablePagination, TableRow, TableFooter, styled, tableCellClasses } from '@mui/material';
import { IProduct } from '../../types/types.ts';
import { useState } from 'react';
import RowComponent from './RowComponent.tsx';

const StyledTableCell = styled(TableCell)(({ theme }) => ({
    [`&.${tableCellClasses.head}`]: {
        backgroundColor: theme.palette.common.white,
        color: theme.palette.common.black,
    },
    [`&.${tableCellClasses.body}`]: {
        fontSize: 14,
    },
}));

type InventoryTableProps = {
    isLoading: boolean;
    filteredProducts: IProduct[];
    setProducts: React.Dispatch<React.SetStateAction<IProduct[]>>
}

const InventoryTable = ({ isLoading, filteredProducts, setProducts }: InventoryTableProps) => {
    const [rowsPerPage, setRowsPerPage] = useState(10);
    const [page, setPage] = useState(0);

    const handleChangeRowsPerPage = (event: React.ChangeEvent<HTMLInputElement>) => {
        setRowsPerPage(Number(event.target.value));
        setPage(0);
    };

    const handleChangePage = (e: React.MouseEvent<HTMLButtonElement, MouseEvent> | null, newPage: number) => {
        e;
        setPage(newPage);
    }

    const rowHeight = (10 - filteredProducts.length) * 55;

    return (
        <TableContainer component='div' id='tableBorder'>
            <Table aria-label='sticky table'>
                <TableHead>
                    <TableRow>
                        <StyledTableCell><b>Code</b></StyledTableCell>
                        <StyledTableCell><b>Name</b></StyledTableCell>
                        <StyledTableCell><b>Category</b></StyledTableCell>
                        <StyledTableCell><b>For Sale</b></StyledTableCell>
                        <StyledTableCell><b>QTY</b></StyledTableCell>
                        <StyledTableCell><b>Actions</b></StyledTableCell>
                    </TableRow>
                </TableHead>
                <TableBody>
                    {filteredProducts.length !== 0
                        ? filteredProducts
                            .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                            .map((product: IProduct) => <RowComponent setProducts={setProducts} product={product} key={product.id} />)
                        : <TableRow>
                            <StyledTableCell>{isLoading ? 'Loading...' : 'No Products'}</StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                        </TableRow>}
                </TableBody>
                <TableFooter>
                    {filteredProducts.length < 10 && window.innerWidth > 768 &&
                        <TableRow sx={{ height: rowHeight }}>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                            <StyledTableCell></StyledTableCell>
                        </TableRow>}
                    <TableRow>
                        <TablePagination
                            rowsPerPageOptions={[10]}
                            count={filteredProducts.length}
                            rowsPerPage={rowsPerPage}
                            page={page}
                            onPageChange={handleChangePage}
                            onRowsPerPageChange={handleChangeRowsPerPage}
                        />
                    </TableRow>
                </TableFooter>
            </Table>
        </TableContainer >
    );
};

export default InventoryTable;   
