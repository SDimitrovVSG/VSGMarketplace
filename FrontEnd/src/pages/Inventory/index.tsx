import { useGetInventoryQuery } from "../../services/productsService";
import { useGetLocationQuery } from "../../services/locationsService.ts";
import { ILocation, IProduct } from "../../types/types.ts";
import { useState, useEffect } from "react";
import AddProductModal from "./AddProductModal.tsx";
import LocationOnIcon from '@mui/icons-material/LocationOn';
import InventoryTable from "./InventoryTable.tsx";
import SearchIcon from "@mui/icons-material/Search";
import AddIcon from "@mui/icons-material/Add";
import {
    Box,
    TextField,
    FormControl,
    InputLabel,
    Select,
    MenuItem,
} from "@mui/material";

const Inventory = () => {
    const { data: locations } = useGetLocationQuery();
    const { data, isLoading } = useGetInventoryQuery();
    const [locationOption, setLocationOption] = useState(0);
    const [products, setProducts] = useState<IProduct[]>([]);
    const [showAddModal, setShowAddModal] = useState(false);
    const [searchString, setSearchString] = useState("");

    useEffect(() => {
        if (data) {
            if (!locationOption) {
                setProducts(data);
            } else {
                const filteredByLocation = data.filter(p => p.locationId === Number(locationOption));
                setProducts(filteredByLocation);
            }
        }
    }, [data, locationOption]);

    const filteredProducts = products.filter((p) =>
        p.fullName.toLowerCase().includes(searchString.toLowerCase())
    );

    return (
        <>
            <AddProductModal setProducts={setProducts} showAddModal={showAddModal} setShowAddModal={setShowAddModal} />
            <main id="inventoryMain">
                <div id="searchSpace">
                    <Box className='searchProduct'>
                        <SearchIcon />
                        <TextField label="Search..." variant="standard"
                            onInput={(e) => setSearchString((e.target as HTMLInputElement).value)}
                        />
                    </Box>
                    <Box className='location' >
                        <LocationOnIcon />
                        <FormControl variant='standard'>
                            <InputLabel>Location</InputLabel>
                            <Select
                                value={locationOption}
                                onChange={(e) => setLocationOption(Number(e.target.value))}
                            >
                                <MenuItem value='0' key='0'>All Locations</MenuItem>
                                {locations?.map((l: ILocation) => (
                                    <MenuItem value={l.id} key={l.id}>
                                        {l.city}
                                    </MenuItem>
                                ))}
                            </Select>
                        </FormControl>
                    </Box>
                    <button id="addButton" onClick={() => setShowAddModal(true)}>
                        <div id='addButtonSpace'>
                            <AddIcon />
                            <span>Add new</span>
                        </div>
                    </button>
                </div>
                {products &&
                    <InventoryTable
                        isLoading={isLoading}
                        setProducts={setProducts}
                        filteredProducts={filteredProducts}
                    />
                }
            </main>
        </>
    );
};

export default Inventory;
