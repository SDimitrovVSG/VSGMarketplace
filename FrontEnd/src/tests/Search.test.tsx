import Inventory from '../pages/Inventory';
import userEvent from '@testing-library/user-event';
import mockServer from 'mock-server';
import { render, screen } from 'test-utils';

describe('Inventory', () => {
    beforeAll(() => mockServer.listen());

    afterEach(() => mockServer.resetHandlers());

    afterAll(() => mockServer.close());
    it('should search and find 4 rows', async () => {
        render(<Inventory />);
        const user = userEvent.setup();
        const search = screen.getByLabelText('Search...');

        await user.type(search, 'HP');

        const rows = await screen.findAllByRole('row');
        expect(rows).toHaveLength(4);
        expect(rows).toMatchSnapshot();
    });
});