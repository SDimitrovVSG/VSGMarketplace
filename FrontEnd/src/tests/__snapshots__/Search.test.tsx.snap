// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inventory should search and find 4 rows 1`] = `
[
  <tr
    class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
  >
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-nmbvi2-MuiTableCell-root"
      scope="col"
    >
      <b>
        Code
      </b>
    </th>
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-nmbvi2-MuiTableCell-root"
      scope="col"
    >
      <b>
        Name
      </b>
    </th>
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-nmbvi2-MuiTableCell-root"
      scope="col"
    >
      <b>
        Category
      </b>
    </th>
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-nmbvi2-MuiTableCell-root"
      scope="col"
    >
      <b>
        For Sale
      </b>
    </th>
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-nmbvi2-MuiTableCell-root"
      scope="col"
    >
      <b>
        QTY
      </b>
    </th>
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-nmbvi2-MuiTableCell-root"
      scope="col"
    >
      <b>
        Actions
      </b>
    </th>
  </tr>,
  <tr
    class="MuiTableRow-root css-8exs75-MuiTableRow-root"
    id="1"
    style="opacity: 1; webkit-transition: opacity 500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <td
      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-132a6sb-MuiTableCell-root"
    >
      234234dfbdfbdf
    </td>
    <td
      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-132a6sb-MuiTableCell-root"
    >
      HP 41
    </td>
    <td
      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-132a6sb-MuiTableCell-root"
    >
      Laptop
    </td>
    <td
      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-132a6sb-MuiTableCell-root"
    >
      12
    </td>
    <td
      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-132a6sb-MuiTableCell-root"
    >
      14
    </td>
    <td
      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-132a6sb-MuiTableCell-root"
    >
      <a
        class="edit"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="EditIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </a>
      <a
        class="delete"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DeleteOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </a>
    </td>
  </tr>,
  <tr
    class="MuiTableRow-root MuiTableRow-footer css-1nd9etx-MuiTableRow-root"
  >
    <td
      class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-1lch3uy-MuiTableCell-root"
    />
    <td
      class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-1lch3uy-MuiTableCell-root"
    />
    <td
      class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-1lch3uy-MuiTableCell-root"
    />
    <td
      class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-1lch3uy-MuiTableCell-root"
    />
    <td
      class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-1lch3uy-MuiTableCell-root"
    />
    <td
      class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-1lch3uy-MuiTableCell-root"
    />
  </tr>,
  <tr
    class="MuiTableRow-root MuiTableRow-footer css-1q1u3t4-MuiTableRow-root"
  >
    <td
      class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium MuiTablePagination-root css-fikjyc-MuiTableCell-root-MuiTablePagination-root"
      colspan="1000"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
      >
        <div
          class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
        />
        <p
          class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
        >
          1â€“1 of 1
        </p>
        <div
          class="MuiTablePagination-actions"
        >
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            title="Go to previous page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowLeftIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </button>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            title="Go to next page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowRightIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </button>
        </div>
      </div>
    </td>
  </tr>,
]
`;
